apps/prairielearn/src/lib/canonical-logger.ts

```diff
@@ -0,0 +1,50 @@
+import { AsyncLocalStorage } from 'node:async_hooks';
+
+import { NextFunction, Request, Response } from 'express';
+
+const als = new AsyncLocalStorage<CanonicalLogger>();
+
+export interface ICanonicalLogger {
+  append(entries: Record<string, any>): void;
+  increment(key: string, value?: number): void;
+  data(): Record<string, any>;
+}
+
+export class CanonicalLogger implements ICanonicalLogger {
```

Comment:
For my own edification, why define the interface first rather than just defining the class directly?

Reply:
This allows me to ensure that the class and the exported `canonicalLogger` have exactly the same interfaces. I wanted the indirection of the exported `canonicalLogger` so that one could just import it anywhere and throw things at it.

An alternative would be to only ever export `getCanonicalLogger`, so then every usage would look like `getCanonicalLogger()?.append(...)`. Do you have any preference here?

Reply:
Makes sense! I like the current way that consumers use `canonicalLogger`. Thanks for explaining!